components/collections/codepointtrie_builder/src/wasm.rs

```diff
@@ -22,58 +21,56 @@ where
     T: TrieValue + Into<u32>,
 {
     // Set up the execution environment with a WasiState
-    let args = &[
-        format!("{}", builder.default_value.into()),
-        format!("{}", builder.error_value.into()),
-        match builder.trie_type {
-            TrieType::Fast => "fast",
-            TrieType::Small => "small",
-        }
-        .to_owned(),
-        format!("{}", std::mem::size_of::<T::ULE>() * 8),
-    ];
     let mut wasi_env = WasiState::new("list_to_ucptrie")
-        .stdin(Box::new(Pipe::new()))
         .stdout(Box::new(Pipe::new()))
-        .args(args)
         .finalize()
         .expect("valid arguments + in-memory filesystem");
 
     // Create the WebAssembly instance with the module and the WasiState
     let import_object = wasi_env.import_object(&MODULE).expect("walid wasm file");
     let instance = Instance::new(&MODULE, &import_object).expect("valid wasm file");
 
-    // To write to the stdin, we need a mutable reference to the pipe
-    //
-    // We access WasiState in a nested scope to ensure we're not holding
-    // the mutex after we need it.
-    {
-        let mut state = wasi_env.state();
-        let wasi_stdin = state
-            .fs
-            .stdin_mut()
-            .expect("valid pipe")
-            .as_mut()
-            .expect("valid pipe");
-        // Write each value to the pipe
-        let CodePointTrieBuilderData::ValuesByCodePoint(values) = builder.data;
-        writeln!(wasi_stdin, "{}", values.len()).expect("valid pipe");
+    let CodePointTrieBuilderData::ValuesByCodePoint(values) = builder.data;
 
-        for value in values {
-            let num: u32 = (*value).into();
-            writeln!(wasi_stdin, "{num}").expect("valid pipe");
-        }
+    // Allocate memory inside wasm and copy values.
+    let memory = instance.exports.get_memory("memory").expect("memory");
+    let malloc = instance
+        .exports
+        .get_native_function::<i32, WasmPtr<u32, Array>>("malloc")
```

Comment:
I think technically this should be
```suggestion
        .get_native_function::<usize, WasmPtr<u32, Array>>("malloc")
```
as `malloc` accepts `size_t`. This will also avoid the cast in line 42.

Reply:
This can't be usize as `FromToNativeWasmType`https://docs.rs/wasmer/2.3.0/wasmer/trait.FromToNativeWasmType.html is not implemented for usize.